test/notifications/display_test.dart

```diff
@@ -111,24 +111,47 @@ void main() {
       required String expectedTagComponent,
     }) {
       final expectedTag = '${data.realmUri}|${data.userId}|$expectedTagComponent';
+      final expectedGroupKey = '${data.realmUri}|${data.userId}';
       final expectedId =
         NotificationDisplayManager.notificationIdAsHashOf(expectedTag);
       const expectedIntentFlags =
         PendingIntentFlag.immutable | PendingIntentFlag.updateCurrent;
-      check(testBinding.androidNotificationHost.takeNotifyCalls()).single
-        ..id.equals(expectedId)
-        ..tag.equals(expectedTag)
-        ..channelId.equals(NotificationChannelManager.kChannelId)
-        ..contentTitle.equals(expectedTitle)
-        ..contentText.equals(data.content)
-        ..color.equals(kZulipBrandColor.value)
-        ..smallIconResourceName.equals('zulip_notification')
-        ..extras.isNull()
-        ..contentIntent.which((it) => it.isNotNull()
-          ..requestCode.equals(expectedId)
-          ..flags.equals(expectedIntentFlags)
-          ..intentPayload.equals(jsonEncode(data.toJson()))
-        );
+
+      check(testBinding.androidNotificationHost.takeNotifyCalls())
+        .deepEquals(<Condition<Object?>>[
+          (it) => it.isA<AndroidNotificationHostApiNotifyCall>()
```

Comment:
Hmm, this `isA` check seems unhelpful. It's statically a fact that each of these items is an `AndroidNotificationHostApiNotifyCall`, so we shouldn't have to repeat that.

How about the version suggested at https://github.com/zulip/zulip-flutter/pull/703#discussion_r1643500420 ? Or the version that was in the previous revision — as that comment said, either of those is fine.

Reply:
The issue is if it's `deepEquals(<Condition<AndroidNotificationHostApiNotifyCall>>[])` it [doesn't recognize](https://github.com/dart-lang/test/blob/329c6dff4dfcb178ba6c7009cf1c1b9215b317aa/pkgs/checks/lib/src/collection_equality.dart#L109) that the child is a `Condition`, instead it goes on with comparing `AndroidNotificationHostApiNotifyCall` with `Condition<AndroidNotificationHostApiNotifyCall>` which obviously fails. This is unlike `containsInOrder` which [correctly checks](https://github.com/dart-lang/test/blob/329c6dff4dfcb178ba6c7009cf1c1b9215b317aa/pkgs/checks/lib/src/extensions/iterable.dart#L102-L106) for the generic type `Condition<T>`.

So, that's why the use of `<Condition<Object?>>` here and then asserting `AndroidNotificationHostApiNotifyCall` in the test.

Reply:
I see. OK, in that case let's stick with the `..length`/`..containsInOrder` version. It's one line longer, but that's worth it in order to make the types make sense, given that the `isA` here doesn't logically belong here as part of what the test is checking (because it'll statically always be true).

I did wonder if there was a subtlety I was missing :slightly_smiling_face: in why that version wasn't equivalent to `deepEquals`: https://github.com/zulip/zulip-flutter/pull/703#discussion_r1643500909 — this definitely counts as such a subtlety.

Reply:
In general if there's something that comes up when you're acting on a review comment and it seems like there's an unexpected obstacle or things go in a different direction, please flag that proactively in the thread (either in reply to the previous comment, or as a new GitHub code-review comment), whether or not you end up going ahead and implementing the new direction.

That's one of the things that helps keep code reviews moving forward smoothly, because it saves the reviewer from having to guess about why your revision of that code came out the way it did.

Reply:
For sure, for this one specifically though (`isA`), I'd looked for any existing use of `deepEquals` in the repo after your initial feedback in https://github.com/zulip/zulip-flutter/pull/703#discussion_r1643500909 and had found [this](https://github.com/zulip/zulip-flutter/blob/6ecdbdc94013b77aea7898638395fa1f3f3fa698/test/widgets/app_test.dart#L34-L36). That gave me the impression that it was an OK change to introduce. :)

Reply:
Cool. Yeah, that's a place where `isA` is needed anyway, because statically it's a list of `Route<dynamic>` but we want this element to specifically be a `WidgetRoute`.

The previous comment that this really goes in a different direction from is the one about where in these checks to put the `AndroidNotificationHostApiNotifyCall` type: https://github.com/zulip/zulip-flutter/pull/703#discussion_r1643500420
So it turned out that the `deepEquals` suggestion was incompatible with that comment. But a contradiction between different review comments I make is the kind of thing I'd very much like to learn about explicitly.